db/revision_cache_lru.go

```diff
@@ -539,22 +536,51 @@ func (rc *LRURevisionCache) revCacheMemoryBasedEviction() {
 
 // performEviction will evict the oldest items in the cache till we are below the memory threshold
 func (rc *LRURevisionCache) performEviction() {
-	rc.lock.Lock()
-	defer rc.lock.Unlock()
-	var numItemsRemoved int64
-	for rc.currMemoryUsage.Value() > rc.memoryCapacity {
-		rc.purgeOldest_()
-		numItemsRemoved++
+	var numItemsRemoved, numBytesRemoved int64
+	rc.lock.Lock() // hold rev cache lock to remove items from cache until we're below memory threshold for the shard
+	// check if we are over memory capacity after holding rev cache mutex (protect against another goroutine evicting whilst waiting for mutex above)
+	if currMemoryUsage := rc.currMemoryUsage.Value(); currMemoryUsage > rc.memoryCapacity {
+		// find amount of bytes needed to evict till below threshold
+		bytesNeededToRemove := currMemoryUsage - rc.memoryCapacity
+		for bytesNeededToRemove > numBytesRemoved {
+			value := rc.lruList.Remove(rc.lruList.Back()).(*revCacheValue)
+			delete(rc.cache, value.key)
+			numItemsRemoved++
+			valueBytes := value.getItemBytes()
+			numBytesRemoved += valueBytes
+		}
 	}
+	rc._decrRevCacheMemoryUsage(-numBytesRemoved) // need update rev cache memory stats before release lock to stop other goroutines evicting based on outdated stats
+	rc.lock.Unlock()                              // release lock after removing items from cache
 	rc.cacheNumItems.Add(-numItemsRemoved)
 }
 
-// updateRevCacheMemoryUsage atomically increases overall memory usage for cache and the actual rev cache objects usage
-func (rc *LRURevisionCache) updateRevCacheMemoryUsage(bytesCount int64) {
+// _decrRevCacheMemoryUsage atomically decreases overall memory usage for cache and the actual rev cache objects usage.
+// You should be holding rev cache lock in using this function to avoid eviction processes over evicting items
+func (rc *LRURevisionCache) _decrRevCacheMemoryUsage(bytesCount int64) {
+	// We need to keep track of the current LRURevisionCache memory usage AND the overall usage of the cache. We need
+	// overall memory usage for the stat added to show rev cache usage plus we need the current rev cache capacity of the
+	// LRURevisionCache object for sharding the rev cache. This way we can perform eviction on per shard basis much like
+	// we do with the number of items capacity eviction
+	if bytesCount > 0 {
+		// function is for decrementing memory usage, so return if incrementing
+		return
+	}
```

Comment:
I like that there's a guard here but I worry about us hitting it silently.
It seems really easy to mistakenly pass in a positive number into a decr function (arguably this is how it should work?)

We should probably be using `AssertfCtx` before the return, which will panic in dev mode, and emit a warning in prod code, since it's a coding error and needs to be addressed.